File: 21-merge-two-sorted-lists/merge-two-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if(list1 == NULL) return list2;
        if(list2 == NULL) return list1;
        ListNode *curr = NULL, *head = NULL;
        while(list1 != NULL && list2 != NULL){
            if(list1 -> val <= list2 -> val){
                if(curr == NULL){
                    curr = list1;
                    head = curr;
                } else {
                    curr -> next = list1;
                    curr = curr -> next;
                }
                list1 = list1 -> next;
            } else {
                if(curr == NULL){
                    curr = list2;
                    head = curr;
                } else {
                    curr -> next = list2;
                    curr = curr -> next;
                }
                list2 = list2 -> next;
            }
        }
        if(list1) curr -> next = list1;
        else curr -> next = list2;
        return head;
    }
};

File: 21-merge-two-sorted-lists/merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if list1 is None:
            return list2
        if list2 is None:
            return list1
        
        head = list1 if (list1.val <= list2.val) else list2

        c1 = list1
        c2 = list2
        curr = None

        while(c1 and c2):
            if c1.val <= c2.val:
                if curr:
                    curr.next = c1
                curr = c1
                c1 = c1.next
            else:
                if curr:
                    curr.next = c2
                curr = c2
                c2 = c2.next
        
        if c1:
            curr.next = c1
        else:
            curr.next = c2
        
        return head
        